### PR TITLE
Introduce Priority field in ServingRuntime

### LINKED ISSUE #3073: Introduce Priority for serving runtimes
/kind feature

**Problem**
- Currently, In auto selection mode, kserve serving runtime does not have a stable way to select a serving runtime if two are more serving runtimes support the same model format. 
- Currently, Kserve [selects the first serving runtime](https://github.com/kserve/kserve/blob/f2f0717a9d6341b6ec9b939bdd324b2c8c507551/pkg/controller/v1beta1/inferenceservice/components/predictor.go#L152) from a list of serving runtimes [sorted  by their respective creation timestamp descending and name ascending order](https://github.com/kserve/kserve/blob/f2f0717a9d6341b6ec9b939bdd324b2c8c507551/pkg/apis/serving/v1beta1/predictor_model.go#L162). So the users have little to no control over the runtime selection in auto select mode.

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- Introduce a new optional field called `Priority` in the model formats section. Through this field we can specify the priority of the serving runtimes for a sepcifice model format.
- The higher the priority value, the higher the precedence
- The priority is valid only if auto select is enabled otherwise the priority is not considered.
- If more than one serving runtime supports the model format and does not provided the priority, then the old way is followed for selecting the runtime.
-  The serving runtime with priority takes precedence over the serving runtime with priority not specified.
- The same model formats with same model version cannot have the same priority.

**Example Yaml**
```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: ClusterServingRuntime
metadata:
  name: kserve-mlserver
spec:
  annotations:
    prometheus.kserve.io/path: /metrics
    prometheus.kserve.io/port: '8080'
  containers:
    - env:
        - name: MLSERVER_MODEL_IMPLEMENTATION
          value: '{{.Labels.modelClass}}'
        - name: MLSERVER_HTTP_PORT
          value: '8080'
        - name: MLSERVER_GRPC_PORT
          value: '9000'
        - name: MODELS_DIR
          value: /mnt/models
      image: docker.io/seldonio/mlserver:1.3.2
      name: kserve-container
      resources:
        limits:
          cpu: '1'
          memory: 2Gi
        requests:
          cpu: '1'
          memory: 2Gi
  protocolVersions:
    - v2
  supportedModelFormats:
    - autoSelect: true
      name: sklearn
      priority: 2
      version: '0'
    - autoSelect: true
      name: xgboost
      priority: 2
      version: '1'
    - autoSelect: true
      name: lightgbm
      priority: 2
      version: '3'
    - autoSelect: true
      name: mlflow
      priority: 1
      version: '1'

```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3073

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add priority field to the supportModelFormats on serving runtime CR, when auto selected is marked for the same supported model format the runtime with higher priority value is selected first.
```

**Dependencies:**
- [x] #3063

### CODE REVIEW

File: python/kserve/docs/V1beta1CustomPredictor.md [hunk 1]
```diff
@@ -11,25 +11,28 @@ Name | Type | Description | Notes
...
-**ephemeral_containers** | [**list[V1EphemeralContainer]**](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1EphemeralContainer.md) | List of ephemeral containers run in this pod. Ephemeral containers may be run in an existing pod to perform user-initiated actions such as debugging. This list cannot be specified when creating a pod, and it cannot be modified by updating the pod spec. In order to add an ephemeral container to an existing pod, use the pod&#39;s ephemeralcontainers subresource. This field is beta-level and available on clusters that haven&#39;t disabled the EphemeralContainers feature gate. | [optional] 
+**ephemeral_containers** | [**list[V1EphemeralContainer]**](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1EphemeralContainer.md) | List of ephemeral containers run in this pod. Ephemeral containers may be run in an existing pod to perform user-initiated actions such as debugging. This list cannot be specified when creating a pod, and it cannot be modified by updating the pod spec. In order to add an ephemeral container to an existing pod, use the pod&#39;s ephemeralcontainers subresource. | [optional]
```
Comment (yuzisun) [inline python/kserve/docs/V1beta1CustomPredictor.md:14]: Can we make these generated code to be a separate PR?

File: Makefile [hunk 2]
```diff
@@ -125,17 +129,9 @@ deploy-helm: manifests
 
 undeploy: kustomize
 	${KUSTOMIZE} build config/default | kubectl delete -f -
-	kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io inferenceservice.serving.kserve.io
```
Comment (yuzisun) [inline Makefile:128]: Why these are deleted ?

Comment (sivanantha321) [reply Makefile:128]: The webhooks are already deleted by the first command. So there is no need to delete it separately.
Screenshot from 2023-07-31 11-20-24

File: python/kserve/docs/V1beta1CustomPredictor.md (hunk 1 above)
Comment (sivanantha321) [reply python/kserve/docs/V1beta1CustomPredictor.md:14]: Are you talking about generated documents ?

File: Makefile [hunk 3]
```diff
@@ -76,6 +76,9 @@ deploy: manifests
 	${KUSTOMIZE} edit remove resource certmanager/certificate.yaml; \
 	else ${KUSTOMIZE} edit add resource certmanager/certificate.yaml; fi;
 	${KUSTOMIZE} build config/default | kubectl apply -f -
+	kubectl wait --for=condition=ready pod -l control-plane=kserve-controller-manager -n kserve --timeout=300s
+	sleep 2
```
Comment (yuzisun) [inline Makefile:80]: Is this sleep for applying the runtime resources ?

File: hack/violation_exceptions.list [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,ModelStatus,ModelCopies
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,ModelStatus,ModelRevisionStates
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PodSpec,DeprecatedServiceAccount
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PredictorExtensionSpec,StorageURI
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PredictorSpec,LightGBM
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PredictorSpec,PyTorch
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PredictorSpec,SKLearn
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,PredictorSpec,XGBoost
-API rule violation: names_match,github.com/kserve/kserve/pkg/apis/serving/v1beta1,StorageSpec,StorageKey
+API rule violation: list_type_missing,./pkg/apis/serving/v1alpha1,BuiltInAdapter,Env
```
Comment (yuzisun) [inline hack/violation_exceptions.list]: I think you are generating from the outside of canonical GOPATH. If you checkout the kserve repo under the `$GOPATH/src/github.com/kserve` then it uses the full package path.

File: pkg/apis/serving/v1alpha1/servingruntime_types.go [hunk 5]
```diff
@@ -36,6 +36,16 @@ type SupportedModelFormat struct {
...
+
+	// +kubebuilder:validation:Minimum=1
+
+	// Priority of this serving runtime for auto selection.
+	// This is used to select the serving runtime if more than one serving runtime supports the same model format.
+	// The value should be greater than zero.  The higher the value, the higher the priority.
+	// Priority is not considered if AutoSelect is either false or not specified.
+	// Priority can be overridden by specifying the runtime in the InferenceService.
+	// +optional
+	Priority *int32 `json:"priority,omitempty"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/servingruntime_types.go:48]: What's the default value of priority or if not specified then it is the lowest priority?

File: pkg/apis/serving/v1beta1/predictor_model.go [hunk 6]
```diff
@@ -199,6 +198,22 @@ func sortClusterServingRuntimeList(runtimes *v1alpha1.ClusterServingRuntimeList)
...
 }
 
+func sortSupportedRuntimeByPriority(runtimes []v1alpha1.SupportedRuntime, modelFormat ModelFormat) {
+	sort.Slice(runtimes, func(i, j int) bool {
+		p1 := runtimes[i].Spec.GetPriority(modelFormat.Name)
+		p2 := runtimes[j].Spec.GetPriority(modelFormat.Name)
+
+		if p1 == nil && p2 == nil {
+			return false
+		} else if p1 == nil && p2 != nil {
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/predictor_model.go]: if priority is not specified it means lower priority right?

File: pkg/controller/v1beta1/inferenceservice/components/predictor.go [hunk 7]
```diff
@@ -150,6 +150,9 @@ func (p *Predictor) Reconcile(isvc *v1beta1.InferenceService) (ctrl.Result, erro
 			}
 			// Get first supporting runtime.
 			sRuntime = runtimes[0].Spec
+			//for _, runtime := range runtimes {
+			//	runtime.Spec.SupportedModelFormats[0].Priority
+			//}
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/predictor.go]: Is this code still needed?

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	for _, runtime := range ExistingRuntimes.Items {
+		if err := validateServingRuntimePriority(&clusterServingRuntime.Spec, &runtime.Spec, clusterServingRuntime.Name, runtime.Name); err != nil {
+			return admission.Denied(fmt.Sprintf("%s in the clusterservingruntimes %s and %s", err.Error(), runtime.Name, clusterServingRuntime.Name))
+		}
+	}
+	return admission.Allowed("")
+}
+
+func validateServingRuntimePriority(newSpec *v1alpha1.ServingRuntimeSpec, existingSpec *v1alpha1.ServingRuntimeSpec, existingRuntimeName string, newRuntimeName string) error {
+	// Skip the runtime if it is disabled or both are not multi model runtime or the stale runtime in the api server
```
Comment (yuzisun) [inline pkg/webhook/admission/servingruntime/servingruntime_webhook.go]: Stale runtime means updating the runtime here?

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+		}
+	}
+	if isTheProtocolSame {
+		for _, existingModelFormat := range existingSpec.SupportedModelFormats {
+			for _, newModelFormat := range newSpec.SupportedModelFormats {
+				// Only validate priority if autoselect is ture
+				if (existingModelFormat.Name == newModelFormat.Name && existingModelFormat.IsAutoSelectEnabled() && newModelFormat.IsAutoSelectEnabled()) &&
+					((existingModelFormat.Version == nil && newModelFormat.Version == nil) ||
+						(existingModelFormat.Version != nil && newModelFormat.Version != nil && *existingModelFormat.Version == *newModelFormat.Version)) {
+					if existingModelFormat.Priority != nil && newModelFormat.Priority != nil && *existingModelFormat.Priority == *newModelFormat.Priority {
```
Comment (yuzisun) [inline pkg/webhook/admission/servingruntime/servingruntime_webhook.go:135]: What if both runtimes' priority is set to nil ?

File: charts/kserve-resources/templates/clusterservingruntimes.yaml [hunk 10]
```diff
@@ -8,10 +8,12 @@ spec:
     prometheus.kserve.io/path: "/metrics"
   supportedModelFormats:
     - name: lightgbm
-      version: "2"
+      version: "3"
```
Comment (yuzisun) [inline charts/kserve-resources/templates/clusterservingruntimes.yaml:11]: Also need to add the validation webhook to the chart?

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+	for _, runtime := range ExistingRuntimes.Items {
+		if err := validateServingRuntimePriority(&clusterServingRuntime.Spec, &runtime.Spec, clusterServingRuntime.Name, runtime.Name); err != nil {
+			return admission.Denied(fmt.Sprintf("%s in the clusterservingruntimes %s and %s", err.Error(), runtime.Name, clusterServingRuntime.Name))
+		}
+	}
+	return admission.Allowed("")
+}
+
+func validateServingRuntimePriority(newSpec *v1alpha1.ServingRuntimeSpec, existingSpec *v1alpha1.ServingRuntimeSpec, existingRuntimeName string, newRuntimeName string) error {
```
Comment (yuzisun) [inline pkg/webhook/admission/servingruntime/servingruntime_webhook.go:117]: Would be nice to add test cases for this validation.

File: Makefile (hunk 3 above)
Comment (sivanantha321) [reply Makefile:80]: The sleep is for servingruntime validating webhook to be ready. Occasionally,  I am getting webhook errors

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go (hunk 9 above)
Comment (sivanantha321) [reply pkg/webhook/admission/servingruntime/servingruntime_webhook.go:135]: Then we won't validate for priority.

File: charts/kserve-resources/templates/clusterservingruntimes.yaml (hunk 10 above)
Comment (sivanantha321) [reply charts/kserve-resources/templates/clusterservingruntimes.yaml:11]: Yes, I forgot.

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go (hunk 8 above)
Comment (sivanantha321) [reply pkg/webhook/admission/servingruntime/servingruntime_webhook.go]: No, The serving runtime is not updated here. In update operations, when we get the servingruntimes from the api server for validation, we also get the same servingruntime which is going to be updated. In that case we have to skip that runtime.

File: pkg/apis/serving/v1beta1/predictor_model.go (hunk 6 above)
Comment (sivanantha321) [reply pkg/apis/serving/v1beta1/predictor_model.go]: Actually, It is affected by the sorting https://github.com/kserve/kserve/blob/32be7464b1a23a59c7980f91175c6c1ff35f4b0a/pkg/apis/serving/v1beta1/predictor_model.go#L100. serving runtime is sorted by created timestamp desc and name asc. So if both of the runtimes does not specify the priority then the order is maintained as it is. If any one of them specifies the priority then that takes precedence.

File: pkg/apis/serving/v1alpha1/servingruntime_types.go (hunk 5 above)
Comment (sivanantha321) [reply pkg/apis/serving/v1alpha1/servingruntime_types.go:48]: There is no default value. For now it is just considered nil.

File: pkg/controller/v1beta1/inferenceservice/components/predictor.go (hunk 7 above)
Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/components/predictor.go]: Will be removed

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go (hunk 11 above)
Comment (sivanantha321) [reply pkg/webhook/admission/servingruntime/servingruntime_webhook.go:117]: Added

File: config/webhook/manifests.yaml [hunk 12]
```diff
@@ -135,3 +135,57 @@ webhooks:
...
+          - CREATE
+          - UPDATE
+        resources:
+          - clusterservingruntimes
+---
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  creationTimestamp: null
+  name: servingruntime.serving.kserve.io
```
Comment (ckadner) [inline config/webhook/manifests.yaml:170]: We will need to rename the validating webhook of the same name in ModelMesh

https://github.com/kserve/kserve/actions/runs/5741094894/job/15561730774#step:10:209

```
helm install kserve charts/kserve-resources/ --wait --timeout 180s

...

Error: INSTALLATION FAILED: 1 error occurred:
	* validatingwebhookconfigurations.admissionregistration.k8s.io "servingruntime.serving.kserve.io" already exists


make: *** [Makefile:128: deploy-helm] Error 1
Error: Process completed with exit code 2.
```

Comment (ckadner) [reply config/webhook/manifests.yaml:170]: #3063

Comment (ckadner) [reply config/webhook/manifests.yaml:170]: PR on the ModelMesh repo: https://github.com/kserve/modelmesh-serving/pull/415

Comment (ckadner) [reply config/webhook/manifests.yaml:170]: Issue on the KServe repo: https://github.com/kserve/kserve/issues/3063
PR on the KServe repo: https://github.com/kserve/kserve/pull/3064
PR on the ModelMesh repo: https://github.com/kserve/modelmesh-serving/pull/415

FYI @Jooho

Comment (ckadner) [reply config/webhook/manifests.yaml:170]: @Jooho

Comment (ckadner) [reply config/webhook/manifests.yaml:170]: @sivanantha321 -- PR #3064 is merged now

File: charts/kserve-resources/templates/clusterservingruntimes.yaml [hunk 13]
```diff
@@ -42,15 +44,19 @@ spec:
     - name: sklearn
       version: "0"
```
Comment (yuzisun) [inline charts/kserve-resources/templates/clusterservingruntimes.yaml:45]: I think mlserver is also supporting version 1 now?

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+			break
+		}
+	}
+	if isTheProtocolSame {
+		for _, existingModelFormat := range existingSpec.SupportedModelFormats {
+			for _, newModelFormat := range newSpec.SupportedModelFormats {
+				// Only validate priority if autoselect is ture
+				if (strings.EqualFold(existingModelFormat.Name, newModelFormat.Name) && existingModelFormat.IsAutoSelectEnabled() && newModelFormat.IsAutoSelectEnabled()) &&
+					((existingModelFormat.Version == nil && newModelFormat.Version == nil) ||
+						(existingModelFormat.Version != nil && newModelFormat.Version != nil && *existingModelFormat.Version == *newModelFormat.Version)) {
```
Comment (yuzisun) [inline pkg/webhook/admission/servingruntime/servingruntime_webhook.go]: I think it is better to extract to a separate function to check whether the two serving runtime support the same model format and version.

File: pkg/apis/serving/v1beta1/predictor_model.go [hunk 15]
```diff
@@ -108,21 +107,26 @@ func (m *ModelSpec) GetSupportingRuntimes(cl client.Client, namespace string, is
...
 	for i := range clusterRuntimes.Items {
 		crt := &clusterRuntimes.Items[i]
 		if !crt.Spec.IsDisabled() && crt.Spec.IsMultiModelRuntime() == isMMS &&
-			m.RuntimeSupportsModel(&crt.Spec) && crt.Spec.IsProtocolVersionSupported(modelProtcolVersion) {
-			srSpecs = append(srSpecs, v1alpha1.SupportedRuntime{Name: crt.GetName(), Spec: crt.Spec})
+			m.RuntimeSupportsModel(&crt.Spec) && crt.Spec.IsProtocolVersionSupported(modelProtocolVersion) {
+			clusterSrSpecs = append(clusterSrSpecs, v1alpha1.SupportedRuntime{Name: crt.GetName(), Spec: crt.Spec})
 		}
 	}
+	log.Info("before sorting", "culsterservingruntimes", clusterSrSpecs)
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/predictor_model.go]: Can remove this debug logging?

File: config/default/kustomization.yaml [hunk 16]
```diff
@@ -159,8 +195,10 @@ patches:
 - path: manager_auth_proxy_patch.yaml
 - path: isvc_mutatingwebhook_cainjection_patch.yaml
 - path: isvc_validatingwebhook_cainjection_patch.yaml
-- path: trainedmodel_mutatingwebhook_cainjection_patch.yaml
+- path: inferencegraph_validatingwebhook_cainjection_patch.yaml
```
Comment (yuzisun) [inline config/default/kustomization.yaml:198]: inference graph validation webhook was not there before?

Comment (sivanantha321) [reply config/default/kustomization.yaml:198]: It was there. It's just the file name is wrong. I just renamed the file.

File: charts/kserve-resources/templates/clusterservingruntimes.yaml (hunk 13 above)
Comment (sivanantha321) [reply charts/kserve-resources/templates/clusterservingruntimes.yaml:45]: Added

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go (hunk 14 above)
Comment (sivanantha321) [reply pkg/webhook/admission/servingruntime/servingruntime_webhook.go]: Done

File: pkg/apis/serving/v1beta1/predictor_model.go (hunk 15 above)
Comment (sivanantha321) [reply pkg/apis/serving/v1beta1/predictor_model.go]: Done

File: pkg/webhook/admission/servingruntime/servingruntime_webhook.go [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	// Only validate for priority if the new serving runtime is not disabled
+	if servingRuntime.Spec.IsDisabled() {
+		return admission.Allowed("")
+	}
+
+	for _, runtime := range ExistingRuntimes.Items {
+		if err := validateServingRuntimePriority(&servingRuntime.Spec, &runtime.Spec, servingRuntime.Name, runtime.Name); err != nil {
```
Comment (yuzisun) [inline pkg/webhook/admission/servingruntime/servingruntime_webhook.go]: @sivanantha321 Generally we do not take the address of the loop variable as it is reused in each iteration, see
https://www.evanjones.ca/go-gotcha-loop-variables.html

**Comment by yuzisun** [comment]
Thanks @sivanantha321 ! Please help add the corresponding priority docs to the serving runtime docs 
https://kserve.github.io/website/0.11/modelserving/servingruntimes/
